### ISSUE TITLE
Entitas in multiplayer Game (using UNet)

### ISSUE THREAD

**Issue by Alexandre-Dhalenne** [sschmid/Entitas-CSharp#802]
Hi,

I would like to know if there is a way to synchronize entities on the host and on the client.
When a client connect to the server, I have overrided OnServerAddPlayer() to execute some code, instantiate the player, etc...
For example, imagine this code on the host : 
```
var myEntity = _gameContext.CreateEntity();
myEntity.AddHealth(100);
```
How can I create the exact same entity on the client , and with data synchronized ? Gameobjects are easy to synchronize with NetworkIdentity and all, but I can't figure it out for entities.
Thanks for your help

**Comment by StormRene** [comment]
You could write reactive systems that are reacting on changes of health and sending messages to your server and vice versa with the framework you prefer. You could probably send some kind of identifier so both computer know which entity is meant and than change the data so both are synced.

**Comment by Alexandre-Dhalenne** [comment]
In fact, my problem is a litlle more complicated.
When a player connect to the host, host creates entity, and one of the component is "View", to deal with the gameobject. The gameobject has a network Identity, so it is synced with the client. Good ! But how can i get the reference to the gameobject on the client to have the same entity  ? 
I don't know if i'm clear, but i do not want to use GameObject.Find(), which takes a lot of resources.

**Comment by StormRene** [comment]
Let me get this straight:
On your server an entity is created, based on a view component a GameObject with a Unity network component is also created. Based on this network component, a GameObject is also created in the client. You want to create a corresponding entity because this network component exists in the client on the GameObject?
I don't think that sounds like a problem of Entitas. You could fire an action component with the NetworkId on the prefab which is created in the client, so that an entity is created which also has it as a component.
Usually you don't build something like this. Usually a GameObject is created in the client due to a change of the entity and not vice versa.

**Comment by sudall** [comment]
Here's a way that I've thought about handling this:

For each component that you want synchronized, you'd do something like this:

```
class HealthController : NetworkBehaviour, IHealthListener, IEventListener<GameEntity> {
    [SyncVar(hook = "OnChange")]
    public int health = 100;

    void Awake() {
        this.GetOrCreateEntity().AddHealth(this.health);
    }    

    void OnChange(int newValue) {
        // when the value changes via UNet on the client...
        if (!isServer) {
            // receive the new value from the server
            this.GetOrCreateEntity().ReplaceHealth(newValue);
        }
    }
    
    void OnHealth(GameEntity entity, int newValue) {
        // when the entity's component on the server changes...
        if (isServer) {
            // send the new value to the client
            this.health = newValue;
        }
    }
    
    public void RegisterEventListeners(GameEntity entity)
    {
        entity.AddHealthListener(this);
    }
}
```

The code might be a bit wrong, but that's the general idea.

`GetOrCreateEntity()` is a monobehaviour extension I've created for accessing an entity that is bound to the game object. My strategy with Entitas is based on the Hybrid Unity ECS approach, which means that entities are created when prefabs are instantiated.

Any thoughts or criticisms?

**Comment by Alexandre-Dhalenne** [comment]
@StormRene I see what you want me to do but I think it will not solve my issue.
Let me summarize : 
- On the host, when a player connects, one method is called (let's say OnServerPlayerConnect() ), which i overrided to customize the spawn of a player. It all works well on the host because it is a client and a server.
- On client, when a player connects, the method is called, but on SERVER side. My playerPrefab is instantiated on the host, everything is well. But, because the prefab has networkidentity and stuff, it is spawned on client too, which should be fine, but there is no method on client to detect when the player is instantiated and to get the gameobject just created. So, when I create, client side, my Entity with my components, I can't put the "Networked" PlayerPrefab into my entity because I don't have anyway to get it. 

@sudall I think your strategy is opposed to what Entitas wants us to do. I think Entitas is meant to create an entity, add components, and act on them (like instantiate a gameobject). 
I don't know if instantiate a prefab with entities is a good way to have a proper ECS architecture.

**Comment by sudall** [comment]
@Alexandre-Dhalenne Fair enough. It's not "proper" ECS, but Unity is so heavily based on game objects that I find this strategy easier and faster to work with. I guess this strategy shoots for compatibility with Unity and other libraries over decoupling from Unity. As a solo dev, I feel like I need to lean on existing tech as much as possible.
